dht.go

```diff
@@ -114,23 +125,28 @@ func New(ctx context.Context, h host.Host, options ...opts.Option) (*IpfsDHT, er
 	dht.enableProviders = cfg.EnableProviders
 	dht.enableValues = cfg.EnableValues
 
+	dht.Validator = cfg.Validator
+	dht.mode = ModeClient
+
+	if !cfg.Client {
```

Comment:
We don't actually determine routability, we _MAY_ determine routability if someone creates a goroutine with the `RoutabilityBasedModeSwitching`.

We effectively have three states which we'd like to configure in application land (e.g. IPFS)
1) ClientOnly: you're a low powered device/leech
2) ServerOnly: you're definitely a server and want to mitigate against any autonat attacks
3) Auto: the standard case

We serve cases 1+2 by using the existing code. However, I'd prefer to either refactor the functional option from `Client(bool)` to `StartState(DHTMode)`, or to start in Client mode by default and solve case 2 by having the application manually call `SetMode(Server)`.

We serve case 3 by just exposing `SetMode(DHTMode)` and giving a utility function that toggles the mode automatically.

Reply:
@Stebalien What do you think ?

Reply:
@aschmahmann 

> we MAY determine routability if someone creates a goroutine with the RoutabilityBasedModeSwitching

What do you mean here ? Is the RoutabilityBasedModeSwitching controlled by a functional option ?

> We serve cases 1+2 by using the existing code. 

I don't think we do. Even if I a start a DHT with the `Client` functional optional configured, you'd still make me a server if you discovered I am publicly reachable via the NAT, no ? Also, I don't think we even have a `ServerOnly` mode to protect against AutoNAT false negatives.

> However, I'd prefer to either refactor the functional option from Client(bool) to StartState(DHTMode), or to start in Client mode by default and solve case 2 by having the application manually call SetMode(Server).

IMO, we should have a `Pin(DHTMode)` option to allow users to pin the DHT as a client/server which would disable Routability based switching or ATLEAST support the `Client`(Client ONLY) option we have right now. correctly. If a user dosen't pin a state, we start as Client & flip to a server when NAT tells us.

Reply:
ping @Stebalien for his thoughts.

Reply:
The idea was to move the smarts out of the DHT itself (i.e., decouple the DHT from these events).

When instantiating the DHT, you'd tell it to be a server or a client. Then, if you _wanted_ to automatically transition, you'd call `RoutabilityBasedModeSwitching` on the DHT. _That_ function would call `SetMode` as necessary to switch the DHT's mode.

---

However, I don't feel strongly about this. If you both agree that the functional option route is simpler, I'd have two options:

* The old (deprecated) `Client(true/false)` option. `Client(false)` would no longer mean _server_, but would mean "not client". (I'm also fine removing this and saying "we're making a major breaking release).
* A new `Mode(client|server|auto)` option. I'd like to stay away from "pin" because we overload that term a bit too much.

Reply:
@Stebalien Having users explicitly call an API to enable some functionally that _should_ ideally be default because most users don't have a strong opinion on it makes it hard for them.

We faced the same problems with the `Bootstrap` API  and then enabled it by default using a functional option/bool and it's working quiet well for us now.

@aschmahmann If you are okay, we should go ahead with the functional option using Sir Steb's suggestions.

Reply:
@Stebalien @aarshkshah1992 this is done in the latest draft. Calling the mode switching externally is the more extensible option which is why I liked it initially, but it may be a little too generic.

With the functional options we don't even have to publicly expose `SetMode` and can keep this internal up until someone has an idea or reason for generic mode switching. If/once they do, we can just expose `SetMode` and give it to them.

Another thing here is that either way my understanding was that we wanted to make the default Client mode and only be server if A) we were explicitly passed in a Server flag B) AutoNAT told us it was ok. As you'll see in the latest update this breaks a whole bunch of tests that assume that either Server is the default or the Client(false) == Server. This was an oversight in the previous version (I left the default as Server mode), but if we really wanted to we could put it back see https://github.com/libp2p/go-libp2p-kad-dht/pull/469#discussion_r388089861